blob/service.go

```diff
@@ -170,40 +171,31 @@ func (s *Service) GetAll(ctx context.Context, height uint64, namespaces []share.
 	var (
 		resultBlobs = make([][]*Blob, len(namespaces))
 		resultErr   = make([]error, len(namespaces))
+		wg          = sync.WaitGroup{}
 	)
-
-	for _, namespace := range namespaces {
-		log.Debugw("performing GetAll request", "namespace", namespace.String(), "height", height)
-	}
-
-	wg := sync.WaitGroup{}
 	for i, namespace := range namespaces {
 		wg.Add(1)
 		go func(i int, namespace share.Namespace) {
+			log.Debugw("retrieving all blobs from", "namespace", namespace.String(), "height", height)
 			defer wg.Done()
+
 			blobs, err := s.getBlobs(ctx, namespace, header)
-			if err != nil {
-				resultErr[i] = fmt.Errorf("getting blobs for namespace(%s): %w", namespace.String(), err)
-				return
+			switch {
+			case err == nil:
+				log.Debugw("receiving blobs", "height", height, "total", len(blobs))
+				resultBlobs[i] = blobs
+			case errors.Is(err, ErrBlobNotFound):
+			default:
+				log.Debug("getting blobs for namespace(%s): %w", namespace.String(), err)
+				resultErr[i] = err
 			}
-
-			log.Debugw("receiving blobs", "height", height, "total", len(blobs))
-			resultBlobs[i] = blobs
 		}(i, namespace)
 	}
 	wg.Wait()
 
-	blobs := make([]*Blob, 0)
-	for _, resBlobs := range resultBlobs {
-		if len(resBlobs) > 0 {
-			blobs = append(blobs, resBlobs...)
-		}
-	}
-
-	if len(blobs) == 0 {
-		resultErr = append(resultErr, ErrBlobNotFound)
-	}
-	return blobs, errors.Join(resultErr...)
+	blobs := slices.Concat(resultBlobs...)
+	err = errors.Join(resultErr...)
+	return blobs, err
```

Comment:
Would information of which blobs are returned and which resulted in error represented in final response? It should be there to ensure user will be able to retry missing blobs if needned.

Reply:
Errors contained namespaces previously and it turned into a mess. The user will have to parse multiple errors in one string which is pretty hard and it is much easier to parse the result blobs and find missing namespaces.

Reply:
>Errors contained namespaces previously and it turned into a mess. The user will have to parse multiple errors in one string which is pretty hard and it is much easier to parse the result blobs and find missing namespaces.

I agree. Aggregating this information into error is a recipe for disaster. 

Ideal solution would be to not request multiple namespaces as described here https://github.com/celestiaorg/celestia-node/issues/3390#issuecomment-2109879641

However if making API to work with single namespace is not feasible right now alternative approach can be used.  To allow parsable errors returned for result should be changed from:
```
[]*Blob, error
```
To
```
[]Response, error

type Response struct {
    []Blob
    Namespace
    error
}
```

Global error would be set only if call failed entirely. This approach will simplify error handling for user by a lot, while also providing granular control on what was retrieved and what was not and need to be retried.

Reply:
I agree that it could solve the problem but in this case, we should rework all our modules to make it consistent.

Reply:
Another we could do is to return `[][]*Blob, []error` with the order matching the order of the namespaces given